Add Guides 2.0 (#90)
hanami/guides: content/v2.0/router/advanced-usage.md

---
title: "Router: Advanced Usage"
order: 30
---

[Refer](https://github.com/hanami/router/blob/unstable/spec/integration/hanami/router/mount_spec.rb)

## Mounting an application

In Hanami, you can mount any rack application in desired route. If we want to do it, you should use the `mount` keyword.

```ruby
# /config/routes.rb

Hanami.application.routes do
  mount Api::App.new, at: "/api"
  mount Backend::App, at: "/backend"
  mount \->(\*) { \[200, {"Content-Length" => "4"}, \["proc"\]\] }, at: "/proc"
  mount \->(\*) { \[200, {"Content-Length" => "8"}, \["trailing"\]\] }, at: "/trailing/"
end
```

#### Mounting To A Path

```ruby
# /config/routes.rb

Hanami.application.routes do
  mount SinatraApp.new, at: "/sinatra"
end
```

All the HTTP requests starting with `/sinatra` will be routed to `SinatraApp`.

#### Mounting On A Subdomain

```ruby
# /config/routes.rb

Hanami.application.routes do

>     mount Blog.new, host: "blog"

Reviewer: ```suggestion
  mount Blog.new, host: "blog"
```

end
```

All the HTTP requests to `http://blog.example.com` will be routed to `Blog`.

<!--   TODO: Update this notice to meet 2.0 behaviour
<p class="notice">
  In development, you will NOT be able to access <code>http://blog.localhost:2300</code>,
  so you should specify a host when running the server:
  <code>bundle exec hanami server --host=lvh.me</code>.
  Then your application can be visited at <code>http://blog.lvh.me:2300</code>
</p>
-->

### RESTful Resources

Hanami has native [REST](http://en.wikipedia.org/wiki/Representational_state_transfer) support.

At the routing level, there are two methods that can be used to declare them: `resources` and `resource`.
The former is for plural resources, the latter for singular ones.

Declaring a resource means to generate **several default routes** with just one line of code.

## RESTful Resources

### Default Routes

```ruby
# /config/routes.rb

Hanami.applicaton.routes do
  resources :books
end

```

It generates:

> <table class="table table-bordered table-striped">

Reviewer: Can we change this to (GitHub-flavored) Markdown tables? https://github.github.com/gfm/#tables-extension-

Author: Sure. I copied it from existing docs, but if other syntax works better, I don't mind.

  <tr>
    <th>Verb</th>
    <th>Path</th>
    <th>Action</th>
    <th>Name</th>
    <th>Named Route</th>
  </tr>
  <tr>
    <td>GET</td>
    <td>/books</td>
    <td>Books::Index</td>
    <td>:index</td>
    <td>:books</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>/books/:id</td>
    <td>Books::Show</td>
    <td>:show</td>
    <td>:book</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>/books/new</td>
    <td>Books::New</td>
    <td>:new</td>
    <td>:new_book</td>
  </tr>
  <tr>
    <td>POST</td>
    <td>/books</td>
    <td>Books::Create</td>
    <td>:create</td>
    <td>:books</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>/books/:id/edit</td>
    <td>Books::Edit</td>
    <td>:edit</td>
    <td>:edit_book</td>
  </tr>
  <tr>
    <td>PATCH</td>
    <td>/books/:id</td>
    <td>Books::Update</td>
    <td>:update</td>
    <td>:book</td>
  </tr>
  <tr>
    <td>DELETE</td>
    <td>/books/:id</td>
    <td>Books::Destroy</td>
    <td>:destroy</td>
    <td>:book</td>
  </tr>
</table>

### Remove Routes

> In case we don"t need all the default routes we can use `:only` and pass one or more action names.

Reviewer: ```suggestion
In case we don't need all the default routes we can use `:only` and pass one or more action names.
```

> We can also black list routes with `:except`.

Reviewer: ```suggestion
We can also exclude routes with `:except`.
```

```ruby
# /config/routes.rb

Hanami.applicaton.routes do
  resources :books, only: [:new, :create, :show]

  # equivalent to

  resources :books, except: [:index, :edit, :update, :destroy]
end
```

### Add Routes

Alongside with default routes we can specify extra routes for single (`member`) or multiple (`collection`) resources.

```ruby
# /config/routes.rb

Hanami.applicaton.routes do
  resources :books do
    member do
      # Generates /books/1/toggle, maps to Books::Toggle, named :toggle_book
      get "toggle"
    end

    collection do
      # Generates /books/search, maps to Books::Search, named :search_books
      get "search"
    end
  end
end
```

### Configure Controller

Imagine we have a controller named `manuscripts`, where we have actions like `Manuscripts::Index`, but we still want to expose those resources as `/books`.
Using the `:controller` option will save our day.

```ruby
# /config/routes.rb

Hanami.applicaton.routes do
  resources :books, controller: "manuscripts"

  # GET /books/1 will route to Manuscripts::Show, etc.
end
```

## RESTful Resource

> You can also use a `resource` helper, to generate routes for singular resource.

Reviewer: ```suggestion
You can also use the `resource` helper, to generate routes for a singular resource.
```

```ruby
# /config/routes.rb

Hanami.application.routes do
  resource :account
end
```

It generates:

<table class="table table-bordered table-striped">
  <tr>
    <th>Verb</th>
    <th>Path</th>
    <th>Action</th>
    <th>Name</th>
    <th>Named Route</th>
  </tr>
  <tr>
    <td>GET</td>
    <td>/account</td>
    <td>Account::Show</td>
    <td>:show</td>
    <td>:account</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>/account/new</td>
    <td>Account::New</td>
    <td>:new</td>
    <td>:new_account</td>
  </tr>
  <tr>
    <td>POST</td>
    <td>/account</td>
    <td>Account::Create</td>
    <td>:create</td>
    <td>:account</td>
  </tr>
  <tr>
    <td>GET</td>
    <td>/account/edit</td>
    <td>Account::Edit</td>
    <td>:edit</td>
    <td>:edit_account</td>
  </tr>
  <tr>
    <td>PATCH</td>
    <td>/account</td>
    <td>Account::Update</td>
    <td>:update</td>
    <td>:account</td>
  </tr>
  <tr>
    <td>DELETE</td>
    <td>/account</td>
    <td>Account::Destroy</td>
    <td>:destroy</td>
    <td>:account</td>
  </tr>
</table>

> ### Remove Routes

Reviewer: There's a lot of duplicated content in this file, from this point onward. Think a couple extra pastes might have gotten saved by accident :)

title: "Router: Advanced Usage"
order: 30
---

[Refer](https://github.com/hanami/router/blob/unstable/spec/integration/hanami/router/mount_spec.rb)

## Mounting an application

> In Hanami, you can mount any rack application in desired route. If we want to do it, you should use the `mount` keyword.

Reviewer: This is in the rack configuration, right? `config.ru` rather than a route files?

Author: oh, is it? I will check, thanks! I know it was in the config in 1.3, but I think I saw it changed in new tests.

Will get back to it! cc: @jodosha

Reviewer 2: Nope, it's the router.

@swilgosz I strongly advice to wrap those lines into the following block:

```ruby
# config/routes.rb
module MyApp
  class ApplicationRoutes < Hanami::Application::Routes
    mount # ...
  end
end
```

Please be aware that this syntax may change, but at least it's crystal clear of what we're talking about: application routes, and not `config.ru`.

---

I would not use "keyword" but "method" instead. Usually keywords are langs constructs.

Author: @jodosha ATM Hanami2APplication template does not inherit from Hanami::Application::Routes, but rather passes a block to the main class

https://github.com/hanami/hanami-2-application-template/blob/main/config/routes.rb
